### PR TITLE
support for test code reuse

### PR DESCRIPTION
**Pull request by lsylvestre**
**about https://github.com/ocaml-sf/learn-ocaml/issues/169  and  https://github.com/ocaml-sf/learn-ocaml/issues/174.**

This PR offers a support for test code reuse.

It allows the programmer to add an optional file called **depend.txt** (in the **exercises** directory) which declares a list of dependencies as follows :

```
../lib/m1.ml   # just a comment
../lib/m1.mli

../lib/m2.ml
```

Thus, these ml and mli files are copied, ciphered and will be loaded in the toplevel during the grading, just before **test.ml**. It is therefore possible to use the modules *M1* and *M2* in **test.ml** and, furthermore, to use *M1* in *M2*.

### CODE REVIEW

File: src/grader/grading.ml [hunk 1]
```diff
@@ -142,6 +142,38 @@ let get_grade
...
+          let files = File.dependencies (access File.depend exo) in
+          List.map (fun f -> let path = File.key f 
+                             and content = decipher f exo in
+                             (path,content)) files 
+        in
+        let ml_files, mli_files = 
+            List.partition (fun (s,_) -> match Filename.extension s with
+                                         | ".ml" -> true
+                                         | ".mli" -> false
+                                         | _ -> failwith "depend.txt (1)") mods
```
Comment (yurug) [inline src/grader/grading.ml]: This error message is not really informative. Can you fix that?

File: src/grader/grading.ml [hunk 2]
```diff
@@ -142,6 +142,38 @@ let get_grade
...
+        let ml_files, mli_files = 
+            List.partition (fun (s,_) -> match Filename.extension s with
+                                         | ".ml" -> true
+                                         | ".mli" -> false
+                                         | _ -> failwith "depend.txt (1)") mods 
+        in
+        let insert_dependencies_in_env (current_path,structure) =
+          let name = String.capitalize_ascii (Filename.(remove_extension (basename current_path))) in
+          handle_error (internal_error [%i"while loading users dependencies"]) @@
+          match List.find_opt (fun (path,_) -> path = current_path ^ "i") mli_files with
```
Comment (yurug) [inline src/grader/grading.ml]: This part can be formulated in a more elegant way by first definition what `signature` is and then make a single call to `use_mod_string`.

File: src/repo/learnocaml_exercise.ml [hunk 3]
```diff
@@ -17,24 +17,28 @@ type t =
...
 let encoding =
   let open Json_encoding in
   conv
-    (fun { id; prelude; template; descr; prepare; test; solution; max_score } ->
-       id, prelude, template, descr, prepare, test, solution, max_score)
-    (fun (id, prelude, template, descr, prepare, test, solution, max_score) ->
-       { id ; prelude ; template ; descr ; prepare ; test ; solution ; max_score })
-    (obj8
+    (fun { id; prelude; template; descr; prepare; test; solution; max_score; depend; dependencies} ->
+       id, prelude, template, descr, prepare, test, solution, max_score,depend, dependencies)
```
Comment (yurug) [inline src/repo/learnocaml_exercise.ml]: Please insert a space after the comma.

File: src/repo/learnocaml_exercise.ml [hunk 4]
```diff
@@ -179,6 +183,29 @@ module File = struct
       update = (fun solution ex -> { ex with solution })
      }
 
+  let depend =
+    { key = "../depend.txt" ; ciphered = false ; (* does'nt work with ciphered = true *)
```
Comment (yurug) [inline src/repo/learnocaml_exercise.ml]: The comment is not correct english and is really cryptic. Please explain to the reader what is at stake here or make it explicit in the code.

File: src/repo/learnocaml_exercise.ml [hunk 5]
```diff
@@ -292,7 +319,10 @@ module File = struct
           read_file prepare ;
           read_file solution ;
           read_file test ;
+          read_file depend ;
           (* read_max_score () *) ] >>= fun () ->
+      join (let dep = try get depend !ex with Missing_file _ -> "" in
```
Comment (yurug) [inline src/repo/learnocaml_exercise.ml]: I am always suspicious when I see an exception being ignored like this. Can you justify this? Shouldn't you at least inform the user? or even stop with an error?

File: src/repo/learnocaml_exercise.ml [hunk 6]
```diff
@@ -335,6 +364,7 @@ module MakeReaderAnddWriter (Concur : Concur) = struct
     let open Concur in
     FileReader.read ~read_field ?id ?decipher () >>= fun ex ->
     try
+      let dep = try field_from_file File.depend ex with File.(Missing_file _) -> "" in
```
Comment (yurug) [inline src/repo/learnocaml_exercise.ml]: This seems redundant with the previous L324.

**Comment by yurug** [comment]
Thank you for this proposal. Can you document this feature in a new file located in `docs`, please?

**Comment by yurug** [comment]
Ping @lsylvestre

**Comment by lsylvestre** [comment]
Hi @yurug,
  I am sorry for the long delay...
Based on your comments and suggestions, I have improved my solution and written a new step in docs/tutorials. Thank you for your review. 

To prevent a warning, I would like to add the following fragment in [translations/fr.po](https://github.com/ocaml-sf/learn-ocaml/blob/master/translations/fr.po#L935):
```
#: src/grader/grading.ml:173,43--80
msgid "while loading user dependencies"
msgstr "lors du chargement des dépendances"
```
but it seems to produce a conflict.

**Comment by yurug** [comment]
Can you give more details about your conflict? What kind of conflicts is it?

File: docs/tutorials/step-8.md [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+- **solution.ml** in a module `Solution` ;
+- the grading modules **Introspection**, **Report** and **Test_lib**.
+
+### Separating the grader code
+
+It is possible to extend this environment by declaring some other user-defined modules in an optional file **depend.txt**, located in the exercise directory.
+
+Each declaration in **depend.txt** is a single line containing the relative path of an *.ml* or *.mli* file. The order of the *.ml* declarations specifies the order in which each module is loaded in the grading environment.
+
+By default each dependency *foo.ml* is isolated in a module *Foo*, which can be constrained by the content of an optional signature file *foo.mli*. Furthermore, we can add an annotation `@included` at the beginning of a file *foo.ml* to denote that all the bindings of *foo.ml* is are evaluated in the toplevel environment (and not in a module *Foo*).
```
Comment (yurug) [inline docs/tutorials/step-8.md:17]: Please use shorter lines in text files (80 columns max.).

Here, who is "we"?

Comment (yurug) [reply docs/tutorials/step-8.md:17]:
```suggestion
By default each dependency *foo.ml* is isolated in a module *Foo*, which can be constrained by the content of an optional signature file *foo.mli*. Furthermore, we can add an annotation `@included` at the beginning of a file *foo.ml* to denote that all the bindings of *foo.ml* are evaluated in the toplevel environment (and not in a module *Foo*). 
```

Comment (yurug) [reply docs/tutorials/step-8.md:17]: You should provide an example, especially to show what is exactly the syntax for annotations.

File: docs/tutorials/step-8.md [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+
+### Separating the grader code
+
+It is possible to extend this environment by declaring some other user-defined modules in an optional file **depend.txt**, located in the exercise directory.
+
+Each declaration in **depend.txt** is a single line containing the relative path of an *.ml* or *.mli* file. The order of the *.ml* declarations specifies the order in which each module is loaded in the grading environment.
+
+By default each dependency *foo.ml* is isolated in a module *Foo*, which can be constrained by the content of an optional signature file *foo.mli*. Furthermore, we can add an annotation `@included` at the beginning of a file *foo.ml* to denote that all the bindings of *foo.ml* is are evaluated in the toplevel environment (and not in a module *Foo*). 
+
+Note that when you edited an *.ml* or *.mli* dependency file which is not defined at the root of your exercice directory, the change will not be applied. In this case a simple shell command `$ touch test.ml` refresh the exercise.
```
Comment (yurug) [inline docs/tutorials/step-8.md:19]:
```suggestion
Dependencies that are not defined at the root of the exercise repository are ignored by the build system: therefore, if you modify them, do not forget to refresh the timestamp of `test.ml` (using `touch` for instance).
```

File: docs/tutorials/step-8.md [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+│   │   ├── test.ml
+│   │   └── tests
+│   │   │   ├── samples.ml
+│   │       ├── add.ml
+│   │       └── odd_even.ml
+│   ├── an-other-exercise
+│   │   ├── depend.txt
+│   │   │ ...
+```
+The exercise **peano** follows the classical format : **prelude.ml**, **prepare.ml**, **solution.ml**, **template.ml** and **test.ml**. It comprises moreover several dependencies (**check.ml**, **samples.ml**, **add.ml** and **odd_even.ml**) which are declared as follow in **depend.txt**:
```
Comment (yurug) [inline docs/tutorials/step-8.md:49]:
```suggestion
The exercise **peano** follows the classical format : **prelude.ml**, **prepare.ml**, **solution.ml**, **template.ml** and **test.ml**. It also includes several dependencies (**check.ml**, **samples.ml**, **add.ml** and **odd_even.ml**) which are declared as follows in **depend.txt**:
```

File: docs/tutorials/step-8.md [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+let test () =
+  Test_lib.test_function_1_against_solution
+    [%ty : peano -> bool ] "odd"
+    [ Z ; S(Z) ; S(S(Z)) ] 
+  @
+  Test_lib.test_function_1_against_solution
+    [%ty : peano -> bool ] "even"
+    [ Z ; S(Z) ; S(S(Z)) ]
+```
+Remember that **Test_lib** internally requires a user-defined sampler `sample_peano : unit -> peano` to generate value of type `peano`. This sampler has to be present in the toplevel environment -- and not in a module -- in order to be found by the introspection primitives during grading. Therefore, we define this sampler in a file starting by the annotation `@included`.
```
Comment (yurug) [inline docs/tutorials/step-8.md:129]:
```suggestion
Remember that **Test_lib** internally requires a user-defined sampler `sample_peano : unit -> peano` to generate value of type `peano`. This sampler has to be present in the toplevel environment -- and not in a module -- in order to be found by the introspection primitives during grading. Therefore, we define this sampler in a file starting with the annotation `@included`.
```

File: docs/tutorials/step-8.md [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+- **samples.ml**:
+```ocaml
+@included
```
Comment (yurug) [inline docs/tutorials/step-8.md:132]: Using the standard syntax:
```
[@@@included]
```
would have the benefits of making this .ml file accepted by the compiler.

**Comment by lsylvestre** [comment]
The conflict was due to my outdated version of the file `translations/fr.po`. It was ~"easy to solve from the GitHub interface". I try again ?

**Comment by lsylvestre** [comment]
The links are dead in the [actual version of learn-ocaml](https://github.com/ocaml-sf/learn-ocaml/blob/master/docs/howto-write-exercises.md)

File: docs/tutorials/step-8.md (hunk 8 above)
Comment (lsylvestre) [reply docs/tutorials/step-8.md]: thank you. It's updated.

File: docs/tutorials/step-8.md (hunk 9 above)
Comment (lsylvestre) [reply docs/tutorials/step-8.md]: thank you. It's updated.

File: docs/tutorials/step-8.md (hunk 10 above)
Comment (lsylvestre) [reply docs/tutorials/step-8.md]: thank you. It's updated.

File: docs/tutorials/step-8.md (hunk 11 above)
Comment (lsylvestre) [reply docs/tutorials/step-8.md]: I have switched to the standard syntax.

Note that the parsing of the annotation is hard-coded [in grading.ml](https://github.com/lsylvestre/learn-ocaml/blob/e99fbc9d9e020d9abe6d641c5534b40b6e548111/src/grader/grading.ml#L167) as a string comparison:
 ```ocaml
match String.trim (String.sub content 0 i) with 
| "[@@@included]" -> ...
```

Is this the right way to do it?

Comment (yurug) [reply docs/tutorials/step-8.md]: That's probably fine.

A cleaner way would be to inspect the module syntax tree but I am not sure if that is easy for you to get this tree at this point of the process.

**Comment by yurug** [comment]
> > Can you give more details about your conflict? What kind of conflicts is it?
> 
> The conflict was due to my outdated version of the file `translations/fr.po`. It is a ~"simple merge conflicts that can be resolve on GitHub using the conflict editor".
> **Do you want me to commit my change in `translations/fr.po` again ?**

Yes please.

File: docs/tutorials/step-8.md (hunk 7 above)
Comment (yurug) [reply docs/tutorials/step-8.md]: No that's fine.

**Comment by lsylvestre** [comment]
> > > Can you give more details about your conflict? What kind of conflicts is it?
> > 
> > 
> > The conflict was due to my outdated version of the file `translations/fr.po`. It is a ~"simple merge conflicts that can be resolve on GitHub using the conflict editor".
> > **Do you want me to commit my change in `translations/fr.po` again ?**
> 
> Yes please.

It is done. There is a conflict as expected.

**Comment by yurug** [comment]
Thanks! We are done here, @lsylvestre, great work!

**Comment by lsylvestre** [comment]
Thank you @yurug, it was my pleasure.

**Pull request merged**